compiler-core/src/build/package_loader.rs

```diff
@@ -1715,3 +1682,143 @@ impl<'a> Inputs<'a> {
         Ok(())
     }
 }
+
+/// A Gleam source file (`.gleam`) and the module name deduced from it
+pub struct GleamFile {
+    pub path: Utf8PathBuf,
+    pub module_name: EcoString,
+}
+
+impl GleamFile {
+    pub fn new(dir: &Utf8Path, path: Utf8PathBuf) -> Self {
+        Self {
+            module_name: Self::module_name(&path, &dir),
+            path,
+        }
+    }
+
+    /// Iterates over Gleam source files (`.gleam`) in a certain directory.
+    /// Symlinks are followed.
+    /// If the there is a .gleam file with a path that would be an
+    /// invalid module name it should not be loaded. For example, if it
+    /// has a uppercase letter in it.
+    pub fn iterate_files_in_directory<'b>(
+        io: &'b impl FileSystemReader,
+        dir: &'b Utf8Path,
+    ) -> impl Iterator<Item = Result<Self, crate::Warning>> + 'b {
+        tracing::trace!("gleam_source_files {:?}", dir);
+        files_with_extension(io, dir, "gleam").map(move |path| {
+            if (Self::is_gleam_path(&path, &dir)) {
+                Ok(Self::new(dir, path))
+            } else {
+                Err(crate::Warning::InvalidSource { path })
+            }
+        })
+    }
+
+    pub fn cache_files(&self, artefact_directory: &Utf8Path) -> CacheFiles {
+        CacheFiles::new(artefact_directory, &self.module_name)
+    }
+
+    fn module_name(path: &Utf8Path, dir: &Utf8Path) -> EcoString {
+        // /path/to/project/_build/default/lib/the_package/src/my/module.gleam
+
+        // my/module.gleam
+        let mut module_path = path
+            .strip_prefix(dir)
+            .expect("Stripping package prefix from module path")
+            .to_path_buf();
+
+        // my/module
+        let _ = module_path.set_extension("");
+
+        // Stringify
+        let name = module_path.to_string();
+
+        // normalise windows paths
+        name.replace("\\", "/").into()
+    }
```

Comment:
This was moved, and made private so that it wouldn't be used with cache files (which caused some extra recompilation because we thought the modules were removed).